### ISSUE TITLE
CollectToIoRank broken in parallel

### ISSUE THREAD

**Issue by blattms** [OPM/ewoms#139]
While it works sequentially (either when compiled with `-DNDEBUG' or with PR #138, it still get an assertion in a parallel run:
```
flow_ebos: /home/mblatt/src/dune/opm/ewoms/ebos/collecttoiorank.hh:127: Ewoms::CollectDataToIORank<GridManager>::DistributeIndexMapping::DistributeIndexMapping(const std::vector<int>&, const std::vector<int>&, Ewoms::CollectDataToIORank<GridManager>::IndexMapType&, Ewoms::CollectDataToIORank<GridManager>::IndexMapStorageType&) [with GridManager = Ewoms::EclCpGridManager<Ewoms::Properties::TTag::EclFlowProblem>; Ewoms::CollectDataToIORank<GridManager>::IndexMapType = std::vector<int>; Ewoms::CollectDataToIORank<GridManager>::IndexMapStorageType = std::vector<std::vector<int> >]: Assertion `checkPosition_.find( id ) == checkPosition_.end()' failed.

Program received signal SIGABRT, Aborted.
0x00007ffff20e1067 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56      ../nptl/sysdeps/unix/sysv/linux/raise.c: Datei oder Verzeichnis nicht gefunden.
(gdb) up
....
(gdb) 
#4  0x0000000000d71727 in Ewoms::CollectDataToIORank<Ewoms::EclCpGridManager<Ewoms::Properties::TTag::EclFlowProblem> >::DistributeIndexMapping::DistributeIndexMapping (this=0x7fffffffd280, 
    globalIndex=std::vector of length 300, capacity 300 = {...}, 
    distributedGlobalIndex=std::vector of length 348, capacity 348 = {...}, 
    localIndexMap=std::vector of length 174, capacity 174 = {...}, 
    indexMaps=std::vector of length 2, capacity 2 = {...})
    at /home/mblatt/src/dune/opm/ewoms/ebos/collecttoiorank.hh:127
127                             assert( checkPosition_.find( id ) == checkPosition_.end() );
(gdb) p checkPosition_ 
$1 = std::set with 1 elements = {[0] = -1}
(gdb) p id
$2 = -1
(gdb) p distributedGlobalIndex
$3 = std::vector of length 348, capacity 348 = {-1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 
  -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, -1, 0, 1, 2, 3, 4, 
  5, 6, 10, 11, 12, 13, 14, 15, 16, 20, 21, 22, 23, 24, 25, 26, 30, 31, 32, 
  33, 34...}
```
This seem two wrong in two ways: There should be no -1 in there and it should probably start with 0 at the first entry.

I have a few questions:
1. This code is run when setting up the ebos simulator. CollectToIoRank is used by the eclipse writer in ewoms. Is that used by flow_ebos? (I wanted to use the debugger to find that out, but it finds to many bugs before that might happen)
2. Why are we reinventing the wheel (over and over)? There is similar (working) code in opm-simulators (ParallelDebugOutput). The only viable reason would be to support other grids but the comments suggest that this is not the case.

**Comment by blattms** [comment]
I might have found one issue. You are [resizing](/OPM/ewoms/blob/284aeba81d2355451ec9ae7bc1d862e0ee011874/ebos/collecttoiorank.hh#L232) an array and than adding entries using [push_back.](/OPM/ewoms/blob/284aeba81d2355451ec9ae7bc1d862e0ee011874/ebos/collecttoiorank.hh#L244). Probably only part of the problem, though

**Comment by andlaus** [comment]
(tested) patches for this are welcome.

**Comment by blattms** [comment]
*ping* I would need an answer to 1-2), I can trust, urgently, to proceed as I might need to refactor the gathering code.

**Comment by andlaus** [comment]
> 1 [...] Is that used by flow_ebos?

no. the flow side deals with output there.

> 2. Why are we reinventing the wheel (over and over)?

does "history" count? that code is quite a bit older than the stuff in opm-simulators and until a few months ago neither code could use the other because there were no dependencies between the modules.

I think that to make flow_ebos run in parallel, it is sufficient if the eWoms output infrastructure is non-functional but does not segfault on creation of these objects.

**Comment by andlaus** [comment]
any news here? `flow_ebos` is probably not affected by this anymore, right?

**Comment by blattms** [comment]
Resolved with mere of #140
